### PR TITLE
[IOPLT-554] Create activation job handler

### PR DESCRIPTION
**Pull request by kin0992**
Depends on #60

#### List of Changes

- Create the handler to create the activations job

#### Motivation and Context

We need a way to handle the activation requests.

#### Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] I have committed only changes relevant to the task.
- [x] I have minimized the number of changes.
- [ ] My changes are about only one task or issue.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: apps/functions-subscription/src/domain/activation-job.ts [hunk 1]
```diff
@@ -27,3 +33,43 @@ export const ActivationJobCodec = t.strict({
 });
 
 export type ActivationJob = t.TypeOf<typeof ActivationJobCodec>;
+
+export type InsertActivationJob = Pick<
```
Comment (kin0992) [inline apps/functions-subscription/src/domain/activation-job.ts:37]:
```suggestion
type InsertActivationJob = Pick<
```
suggestion(non-blocking): it's not used somewhere else

File: apps/functions-subscription/src/main.ts [hunk 2]
```diff
@@ -94,14 +101,22 @@ app.http('createSubscription', {
...
   handler: makeGetSubscriptionHandler(env),
-  route: '/trials/{trialId}/subscriptions/{userId}',
+  route: 'trials/{trialId}/subscriptions/{userId}',
+});
+
+app.http('createActivationJob', {
+  methods: ['POST'],
+  authLevel: 'function',
+  handler: makeCreateActivationJobHandler(env),
+  // TODO: Change this route accordingly to the OpenAPI. PR#60
```
Comment (kin0992) [inline apps/functions-subscription/src/main.ts:118]:
```suggestion
```

File: apps/functions-subscription/src/main.ts [hunk 3]
```diff
@@ -94,14 +101,22 @@ app.http('createSubscription', {
...
+  route: 'trials/{trialId}/activations',
```
Comment (kin0992) [inline apps/functions-subscription/src/main.ts:119]:
```suggestion
  route: 'trials/{trialId}/activation-jobs',
```

File: apps/functions-subscription/src/adapters/azure/functions/insert-activation-job.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+          usersToActivate: usersToActivate as NonNegativeNumber,
+        }),
+    ),
+    RTE.map(flow(H.successJson, H.withStatusCode(202))),
+    RTE.mapLeft(toHttpProblemJson),
+    RTE.orElseW(RTE.of),
+  ),
+);
+
+export const makeCreateActivationJobHandler = httpAzureFunction(
```
Comment (datalek) [inline apps/functions-subscription/src/adapters/azure/functions/insert-activation-job.ts:41]: **suggestion**: Rename to `makePostActivationJobHandler`

File: apps/functions-subscription/src/domain/__tests__/mocks.ts [hunk 5]
```diff
@@ -1,27 +1,17 @@
...
-    subscriptionHistoryReader: mock<SubscriptionHistoryReader>(),
-    subscriptionHistoryWriter: mock<SubscriptionHistoryWriter>(),
-    activationRequestRepository: mock<ActivationRequestRepository>(),
-    eventWriter: mock<EventWriter>(),
-    clock: mock<Clock>(),
-    hashFn: mockFn<HashFn>(),
-    monotonicIdFn: mockFn<MonotonicIdFn>(),
-  };
-};
+export const makeTestEnv = () => ({
```
Comment (datalek) [inline apps/functions-subscription/src/domain/__tests__/mocks.ts:4]: **note**: These changes are not related to this task

File: apps/functions-subscription/src/domain/activation-job.ts [hunk 6]
```diff
@@ -27,3 +33,40 @@ export const ActivationJobCodec = t.strict({
...
+  );
+
+export const insertActivationJob = (insertActivationJob: InsertActivationJob) =>
+  pipe(
+    RTE.ask<Pick<Capabilities, 'activationJobWriter'>>(),
+    RTE.apSW('activationJob', makeActivationJob(insertActivationJob)),
+    RTE.flatMapTaskEither(({ activationJob, activationJobWriter }) =>
+      activationJobWriter.insert(activationJob),
+    ),
+  );
```
Comment (datalek) [inline apps/functions-subscription/src/domain/activation-job.ts:72]: **note**: We can move it under use-case folder if we prefer

File: apps/functions-subscription/src/adapters/azure/cosmosdb/activation-job.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+import * as E from 'fp-ts/lib/Either';
+import * as TE from 'fp-ts/lib/TaskEither';
+import { Database } from '@azure/cosmos';
+import { ActivationJobWriter } from '../../../domain/activation-job';
+import { cosmosErrorToDomainError } from './errors';
+
+export const makeActivationJobCosmosContainer = (
+  db: Database,
+): ActivationJobWriter => {
+  const container = db.container('activations');
```
Comment (datalek) [inline apps/functions-subscription/src/adapters/azure/cosmosdb/activation-job.ts:11]: **note**: At the moment it is hardcoded but we can refactor and require as parameter of `makeActivationJobCosmosContainer`

Comment (kin0992) [reply apps/functions-subscription/src/adapters/azure/cosmosdb/activation-job.ts:11]: I'd do in a separate PR

**Review by datalek — APPROVED** [review]
